diff -ur uml-utilities-20070815/tunctl/tunctl.c uml-utilities-20070815-pointopoint/tunctl/tunctl.c
--- uml-utilities-20070815/tunctl/tunctl.c	2007-08-16 00:10:03.000000000 +0900
+++ uml-utilities-20070815-pointopoint/tunctl/tunctl.c	2009-11-29 04:14:57.000000000 +0900
@@ -22,7 +22,7 @@
 static void Usage(char *name)
 {
   fprintf(stderr, "Create: %s [-b] [-u owner] [-g group] [-t device-name] "
-	  "[-f tun-clone-device]\n", name);
+	  "[-f tun-clone-device] [-m tun-mode]\n", name);
   fprintf(stderr, "Delete: %s -d device-name [-f tun-clone-device]\n\n",
 	  name);
   fprintf(stderr, "The default tun clone device is /dev/net/tun - some systems"
@@ -38,10 +38,10 @@
   struct group *gr;
   uid_t owner = -1;
   gid_t group = -1;
-  int tap_fd, opt, delete = 0, brief = 0;
+  int tap_fd, opt, delete = 0, brief = 0, layer = 2;
   char *tun = "", *file = "/dev/net/tun", *name = argv[0], *end;
 
-  while((opt = getopt(argc, argv, "bd:f:t:u:g:")) > 0){
+  while((opt = getopt(argc, argv, "bd:f:t:u:g:m:")) > 0){
     switch(opt) {
       case 'b':
         brief = 1;
@@ -83,6 +83,20 @@
       case 't':
         tun = optarg;
         break;
+
+      case 'm':
+	if(!strcmp(optarg, "ethernet")){
+	  layer = 2;
+	}else if (!strcmp(optarg, "pointopoint")){
+	  layer = 3;
+	}else{
+	  fprintf(stderr,
+		  "'%s' is not a tunnel mode (use 'ethernet' or 'pointopoint').\n",
+		  optarg);
+	  Usage(name);
+	}
+	break;
+
       case 'h':
       default:
         Usage(name);
@@ -103,7 +117,7 @@
 
   memset(&ifr, 0, sizeof(ifr));
 
-  ifr.ifr_flags = IFF_TAP | IFF_NO_PI;
+  ifr.ifr_flags = (layer == 2 ? IFF_TAP : IFF_TUN) | IFF_NO_PI;
   strncpy(ifr.ifr_name, tun, sizeof(ifr.ifr_name) - 1);
   if(ioctl(tap_fd, TUNSETIFF, (void *) &ifr) < 0){
     perror("TUNSETIFF");
